quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs

```diff
@@ -161,6 +186,74 @@ impl MappingLeaf {
     }
 }
 
+fn convert_json_val(json_type: JsonType, json_val: JsonValue) -> Option<JsonValue> {
+    match json_val {
+        JsonValue::Number(_) => match json_type {
+            JsonType::Number => Some(json_val),
+            JsonType::Object | JsonType::String => None,
+        },
+        JsonValue::String(_) => match json_type {
+            JsonType::Number | JsonType::Object => None,
+            JsonType::String => Some(json_val),
+        },
+        JsonValue::Object(_) => match json_type {
+            JsonType::Number | JsonType::String => None,
+            JsonType::Object => Some(json_val),
+        },
+        JsonValue::Bool(_) | JsonValue::Null | JsonValue::Array(_) => None,
+    }
+}
+
+fn extract_json_val(
+    json_type: JsonType,
+    named_doc: &mut BTreeMap<String, Vec<JsonValue>>,
+    field_path: &[&str],
+    cardinality: Cardinality,
+) -> Option<JsonValue> {
+    let full_path = field_path.join(".");
+    let vals = named_doc.remove(&full_path)?;
+    match cardinality {
+        Cardinality::SingleValue => {
+            if vals.len() > 1 {
+                debug!(json_path=?field_path, "Received multiple cardinality for a single valued field. Droppping values upon conversion.");
+            }
+            let single_value = vals.into_iter().next()?;
+            convert_json_val(json_type, single_value)
+        }
+        Cardinality::MultiValues => Some(JsonValue::Array(
+            vals.into_iter()
+                .flat_map(|json_value| convert_json_val(json_type, json_value))
+                .collect(),
+        )),
+    }
+}
+
+fn insert_json_val(
+    field_path: &[&str],
+    json_val: JsonValue,
+    mut doc_json: &mut serde_json::Map<String, serde_json::Value>,
+) {
+    assert!(!field_path.is_empty());
+    for field_name in &field_path[..field_path.len() - 1] {
```

Comment:
[split_last](https://doc.rust-lang.org/std/primitive.slice.html#method.split_last) should make the intent clearer

Reply:
agreed!

Reply:
Actually I did not know split_last. very nice!